chore: Use pickle by default when saving Scikit-learn models

Ark-kun opened the pull request:
CloudPickle officially states: "Using cloudpickle for long-term object storage is not supported and strongly discouraged."

Fixes https://github.com/mlflow/mlflow/issues/5419
Fixes https://github.com/mlflow/mlflow/issues/5420

Signed-off-by: Alexey Volkov <alexey.volkov@ark-kun.com>

## What changes are proposed in this pull request?

(Please fill in changes proposed in this fix)

## How is this patch tested?

(Details)

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

## Release Notes

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->
<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

Ark-kun committed:
chore: Use pickle by default when saving Scikit-learn models

CloudPickle officially states: "Using cloudpickle for long-term object storage is not supported and strongly discouraged."

Fixes https://github.com/mlflow/mlflow/issues/5419
Fixes https://github.com/mlflow/mlflow/issues/5420

Signed-off-by: Alexey Volkov <alexey.volkov@ark-kun.com>

dbczumar reviewed (changes requested):
@Ark-kun We use cloudpickle as the standard serialization format for MLflow Models (e.g. models saved by `mlflow.sklearn`) because it has several nice features, such as the ability to serialize objects defined in the `__main__` scope (which is the scope used by notebook environments, such as Jupyter) and the ability to serialize object types not supported by pickle.

We reason that cloudpickle is acceptable for long-term persistence with MLflow because we also record the exact software environment that was used to train the model and provide tools for restoring it.

From what I can tell, cloudpickle uses the default Python pickle protocol: https://github.com/cloudpipe/cloudpickle/blob/1bf52774afca43e3878b5f8256f437484f218ea2/cloudpickle/cloudpickle.py#L83, rather than a custom one.

Please let us know how we can satisfy your use cases while preserving the benefits of serializing models dependent on objects defined in `__main__` scope.

jinzhang21 commented:
I agree with @dbczumar that the rigid requirements on runtime environment is feature not a bug in the context of model reproducibility.

WeichenXu123 commented:
@dbczumar @jinzhang21 
So load model (unpickling) might fail when using different python version or different cloudpickle version ?

Shall we make https://github.com/mlflow/mlflow/blob/ddfcccc631b56cae8083dbb00510d1215a1deded/mlflow/pyfunc/__init__.py#L705 show the warning by default ? If different python version might cause loading failed, then it might be a "fatal" issue we need emit warning.

jinzhang21 commented:
Showing the Python version difference warning by default SGTM. @WeichenXu123

Ark-kun commented:
>We reason that cloudpickle is acceptable for long-term persistence with MLflow because we also record the exact software environment that was used to train the model and provide tools for restoring it.

There are several issues here:
1. Lock-in - Can the user take the model logged by MLflow and use it outside MLflow if needed. I think MLflow should try to be more compatible than native model saving methods, not less. Currently, if I use Scikit-Learn native way to save the model, I can load it under any python version. But I cannot load a Scikit-learn model saved by MLflow.
2. MLflow does not use the model virtual environment in some cases. Most people use MLflow in a notebook. Not in a container. And when you use MLflow in a notebook, you cannot even load MLflow model: `mlflow.pyfunc.load_model` just crashes since it cannot unpickle the model. See https://github.com/mlflow/mlflow/issues/5419

>From what I can tell, cloudpickle uses the default Python pickle protocol: https://github.com/cloudpipe/cloudpickle/blob/1bf52774afca43e3878b5f8256f437484f218ea2/cloudpickle/cloudpickle.py#L83, rather than a custom one.

I'm not sure about that. Python's default pickle protocol is `pickle.DEFAULT_PROTOCOL` while `cloudpickle` uses `pickle.HIGHEST_PROTOCOL`.

dbczumar commented:
> > We reason that cloudpickle is acceptable for long-term persistence with MLflow because we also record the exact software environment that was used to train the model and provide tools for restoring it.
> 
> There are several issues here:
> 
> 1. Lock-in - Can the user take the model logged by MLflow and use it outside MLflow if needed. I think MLflow should try to be more compatible than native model saving methods, not less. Currently, if I use Scikit-Learn native way to save the model, I can load it under any python version. But I cannot load a Scikit-learn model saved by MLflow.
> 2. MLflow does not use the model virtual environment in some cases. Most people use MLflow in a notebook. Not in a container. And when you use MLflow in a notebook, you cannot even load MLflow model: `mlflow.pyfunc.load_model` just crashes since it cannot unpickle the model. See [[BUG] Models saved under Python 3.8 cannot be loaded under Python 3.7 #5419](https://github.com/mlflow/mlflow/issues/5419)
> 
> > From what I can tell, cloudpickle uses the default Python pickle protocol: https://github.com/cloudpipe/cloudpickle/blob/1bf52774afca43e3878b5f8256f437484f218ea2/cloudpickle/cloudpickle.py#L83, rather than a custom one.
> 
> I'm not sure about that. Python's default pickle protocol is `pickle.DEFAULT_PROTOCOL` while `cloudpickle` uses `pickle.HIGHEST_PROTOCOL`.

@Ark-kun Understood. There's definitely a tradeoff between functionality and cross-version compatibility here. By using MLflow with cloudpickle, users can easily develop, serialize, and deploy scikit-learn models with custom transformers from notebook environments. By using MLflow with pickle, they cannot. Right now, we favor the increased functionality over compatibility across Python versions. We recognize that some users have different preferences and requirements, which is why we expose the `serialization_format` argument.

Regarding (2), we are actively working on tools for restoring the original model environment in all contexts where MLflow is available. We are excited about this new feature and hope to introduce it in an upcoming release. cc @WeichenXu123.

Ark-kun commented:
Couple of additional points regarding the context of this PR:

>which is why we expose the serialization_format argument.

At the time this PR was created, the `serialization_format` was not exposed on `autolog` methods. I've fixed that in https://github.com/mlflow/mlflow/pull/5329

>By using MLflow with cloudpickle, users can easily develop, serialize, and deploy scikit-learn models with custom transformers from notebook environments.

I understand that there are cases where `cloudpickle` is the only way and should be used. However I think there are cases where `cloudpickle` is not needed and not using it would increase compatibility. (I admit that this PR makes a crude change which stops using `cloudpickle` even where needed.). Since `cloudpickle` is an extension of `pickle`, not a different format, the biggest issue is really with the pickle protocol it used. I've fixed that issue in https://github.com/mlflow/mlflow/pull/5422

So, for my use case, the issue is essentially solved (by https://github.com/mlflow/mlflow/pull/5422) for users of the next MLflow version (the previously logged models will still be problematic).

P.S. I used `cloudpickle` before in my [KFP SDK](https://github.com/Ark-kun/kfp_sdk_components/blob/a1bd6b35d247255aaa6a3126a9891a0012878a65/components/_python_op.py#L107) project to preserve dependencies of a python function. However due to unstable ABI and incompatibilities (segfaults) even between patch versions I've reverted back to source-based code saving.

dbczumar commented:
Closing this after our prior discussion. Thanks for the context, @Ark-kun; we may gravitate towards source-based code saving in the future, if cloudpickle presents issues.
